### PR TITLE
[BUGFIX] Fix copying of CheckpointConfig for substitution and printing purposes

### PR DESCRIPTION
**Pull request by alexsherstinsky**
### Scope
This work fixes a bug in `CheckpointConfig` serialization, specifically the loss of `batch_data` as the side effect.

The implementation consists of:
* Overriding `CheckpointConfigSchema.dump()`,
* Overriding `CheckpointConfig.__deepcopy__()`, and
* Implementing `CheckpointConfig.__repr__()` and `CheckpointResult.__repr__()` methods

in the way that:
1. Extracts the `batch_data` values from the `CheckpointConfiguration` and saving them.
2. Deleting the `batch_data` values from the `CheckpointConfiguration` and cloning/serializing the resulting "safe" configuration object.
3. Restoring the `batch_data` values into the cloned and/or original configuration object.

Additional clean up consists of:
* Fixing test(s) that returned incorrect results due to the previous implementation.
* Making the `get_runtime_batch_request()` Checkpoint utility more elegant.

As the result of the above fixes, the impacted modules became simpler, and the amount of code has been reduced.

Please annotate your PR title to describe what the PR does, then give a brief bulleted description of your PR below. PR titles should begin with [BUGFIX], [FEATURE],  [DOCS], or [MAINTENANCE]. If a new feature introduces breaking changes for the Great Expectations API or configuration files, please also add [BREAKING]. You can read about the tags in our [contributor checklist](https://docs.greatexpectations.io/en/latest/contributing/contribution_checklist.html).

Changes proposed in this pull request:
- JIRA: ANT-20/ANT-22/ANT-57
-
-

After submitting your PR, CI checks will run and @tiny-tim-bot will check for your CLA signature.

For a PR with nontrivial changes, we review with both design-centric and code-centric lenses.

In a design review, we aim to ensure that the PR is consistent with our relationship to the open source community, with our software architecture and abstractions, and with our users' needs and expectations. That review often starts well before a PR, for example in github issues or slack, so please link to relevant conversations in notes below to help reviewers understand and approve your PR more quickly (e.g. `closes #123`).

Previous Design Review notes:
-
-
-

### Definition of Done
Please delete options that are not relevant.

- [x] My code follows the Great Expectations [style guide](https://docs.greatexpectations.io/en/latest/contributing/style_guide.html?highlight=style%20guide)
- [x] I have performed a [self-review](https://docs.greatexpectations.io/en/latest/contributing/contribution_checklist.html?highlight=checklist) of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added [unit tests](https://docs.greatexpectations.io/en/latest/contributing/testing.html#contributing-testing-writing-unit-tests) where applicable and made sure that new and existing tests are passing.
- [x] I have run any local integration tests and made sure that nothing is broken.

Thank you for submitting!

### CODE REVIEW

File: great_expectations/checkpoint/checkpoint.py [hunk 1]
```diff
@@ -377,13 +327,13 @@ def run(
             "template_name": template_name,
             "run_name_template": run_name_template,
             "expectation_suite_name": expectation_suite_name,
-            "expectation_suite_ge_cloud_id": expectation_suite_ge_cloud_id,
```
Comment (donaldheppner) [inline great_expectations/checkpoint/checkpoint.py:380]: why was this line removed?

File: great_expectations/checkpoint/types/checkpoint_result.py [hunk 2]
```diff
@@ -11,6 +10,9 @@
 from great_expectations.data_context.types.base import (
     CheckpointConfig,
     CheckpointConfigSchema,
+    delete_runtime_parameters_batch_data_references_from_config,
```
Comment (donaldheppner) [inline great_expectations/checkpoint/types/checkpoint_result.py]: could/should these not be methods on the config object?

File: great_expectations/checkpoint/types/checkpoint_result.py [hunk 3]
```diff
@@ -284,97 +286,24 @@ def _list_validation_statistics(self) -> Dict[ValidationResultIdentifier, dict]:
...
-                        serializeable_dict["checkpoint_config"]["batch_request"][
-                            "runtime_parameters"
-                        ]["batch_data"]
-                    )
-                )
-
+        serializeable_dict: dict = self.to_json_dict()
+        batch_data_references: Tuple[
+            Optional[Any], Optional[List[Any]]
+        ] = get_runtime_parameters_batch_data_references_from_config(
```
Comment (donaldheppner) [inline great_expectations/checkpoint/types/checkpoint_result.py:297]: per the above comment re: these imports, rather than importing functions that work against config, consider instantiating a config object with the necessary methods / understanding of itself

File: great_expectations/checkpoint/util.py [hunk 4]
```diff
@@ -207,26 +207,11 @@ def send_email(
 # TODO: <Alex>A common utility function should be factored out from DataContext.get_batch_list() for any purpose.</Alex>
```
Comment (donaldheppner) [inline great_expectations/checkpoint/util.py:208]: Is this TODO still valid?

File: great_expectations/data_context/types/base.py [hunk 5]
```diff
@@ -2269,82 +2296,308 @@ def get_updated_action_list(
...
+                    )
+                else:
+                    validations_batch_data_list.append(None)
+
+    return default_batch_data, validations_batch_data_list
 
-        if (batch_data is not None) and (
-            self.batch_request.get("runtime_parameters") is not None
+
+def delete_runtime_parameters_batch_data_references_from_config(
```
Comment (NathanFarmer) [inline great_expectations/data_context/types/base.py]: ❤️

File: great_expectations/data_context/types/base.py [hunk 6]
```diff
@@ -2269,82 +2296,308 @@ def get_updated_action_list(
...
+                    val.get("batch_request") is not None
+                    and val["batch_request"].get("runtime_parameters") is not None
+                    and "batch_data"
+                    in val["batch_request"]["runtime_parameters"]
+                    is not None
+                ):
+                    val["batch_request"]["runtime_parameters"].pop("batch_data")
+
+
+def restore_runtime_parameters_batch_data_references_into_config(
```
Comment (NathanFarmer) [inline great_expectations/data_context/types/base.py]: ❤️

File: great_expectations/data_context/types/base.py [hunk 7]
```diff
@@ -2269,82 +2296,308 @@ def get_updated_action_list(
...
+            setattr(result, key, value_copy)
+
+        restore_runtime_parameters_batch_data_references_into_config(
+            config=self, batch_data_references=batch_data_references
+        )
+        restore_runtime_parameters_batch_data_references_into_config(
+            config=result, batch_data_references=batch_data_references
+        )
+
+        return result
```
Comment (NathanFarmer) [inline great_expectations/data_context/types/base.py]: Can a custom `__deepcopy__` not be avoided here? Why isn't this something we can do where required, like we are doing in other places using these new functions in this PR?

File: tests/checkpoint/test_checkpoint.py [hunk 8]
```diff
@@ -2337,6 +2337,7 @@ def test_newstyle_checkpoint_config_substitution_nested(
         CheckpointConfig(
             name="my_nested_checkpoint",
             config_version=1,
+            template_name="my_nested_checkpoint_template_3",
```
Comment (NathanFarmer) [inline tests/checkpoint/test_checkpoint.py:2340]: Is there a new test to be added for the bug you referenced?

File: great_expectations/checkpoint/types/checkpoint_result.py [hunk 9]
```diff
@@ -284,97 +286,24 @@ def _list_validation_statistics(self) -> Dict[ValidationResultIdentifier, dict]:
...
-        elif batch_data is not None:
-            self.checkpoint_config["batch_request"]["runtime_parameters"][
-                "batch_data"
-            ] = json_dict["checkpoint_config"]["batch_request"]["runtime_parameters"][
-                "batch_data"
-            ] = batch_data
-
-        return json_dict
+    def to_json_dict(self) -> dict:
+        return checkpointResultSchema.dump(self)
```
Comment (NathanFarmer) [inline great_expectations/checkpoint/types/checkpoint_result.py:292]: ❤️

**Review by NathanFarmer — CHANGES_REQUESTED** [review]
Just a few questions!

File: great_expectations/checkpoint/checkpoint.py (hunk 1 above)
Comment (anthonyburdi) [reply great_expectations/checkpoint/checkpoint.py:380]: Looks like it was just moved to the bottom (new line 336)

Comment (alexsherstinsky) [reply great_expectations/checkpoint/checkpoint.py:380]: @donaldheppner I simply moved it to a "more logical place" congruent with the order these arguments were supplied on input to the method.

File: great_expectations/checkpoint/types/checkpoint_result.py (hunk 2 above)
Comment (alexsherstinsky) [reply great_expectations/checkpoint/types/checkpoint_result.py]: @donaldheppner In fact, even better: in the next commit, they will be moved to `batch.py` core module.

File: great_expectations/checkpoint/types/checkpoint_result.py (hunk 3 above)
Comment (alexsherstinsky) [reply great_expectations/checkpoint/types/checkpoint_result.py:297]: @donaldheppner These are static utility methods -- they will be moved to `batch.py` (in the next commit).

**Review by anthonyburdi — COMMENTED** [review]
Looks good to me aside from resolving Don and Nathan's comments :)

File: great_expectations/checkpoint/util.py (hunk 4 above)
Comment (alexsherstinsky) [reply great_expectations/checkpoint/util.py:208]: @donaldheppner Unfortunately, this comment is still valid -- we have not had a chance to do the work.  I would like to leave it for now and remove only after we had a chance to go through the backlog formally.  Thank you.

File: great_expectations/data_context/types/base.py (hunk 7 above)
Comment (alexsherstinsky) [reply great_expectations/data_context/types/base.py]: @NathanFarmer This custom `__deepcoppy__` enables us to simplify / delete a lot of code.  It fits, because it "knows" how to handle the copying for the (complex) `CheckpointConfig` object.

File: tests/checkpoint/test_checkpoint.py (hunk 8 above)
Comment (alexsherstinsky) [reply tests/checkpoint/test_checkpoint.py:2340]: @NathanFarmer This test has effectively been added by repairing some existing tests.  In addition, the functionality will be fully manifested in a near-future PR that focuses on `checkpoint.run` payload instrumentation, which is how the bug was detected in the first place (the `batch_data` values were empty in the `runtime_parameter` dictionary).  Thanks.

File: great_expectations/core/batch.py [hunk 10]
```diff
@@ -361,8 +361,8 @@ def _validate_runtime_batch_request_specific_init_parameters(
     ):
         if not (runtime_parameters and (isinstance(runtime_parameters, dict))):
             raise TypeError(
-                f"""The type for runtime_parameters must be a dict object.
-                The type given is "{str(type(runtime_parameters))}", which is illegal."""
+                f"""The runtime_parameters must be a non-empty dict object.
```
Comment (Shinnnyshinshin) [inline great_expectations/core/batch.py:407]: Very helpful error message. Thank you for this

File: great_expectations/data_context/types/base.py (hunk 6 above)
Comment (Shinnnyshinshin) [reply great_expectations/data_context/types/base.py]: ❤

File: great_expectations/data_context/types/base.py (hunk 5 above)
Comment (Shinnnyshinshin) [reply great_expectations/data_context/types/base.py]: ❤

File: great_expectations/checkpoint/types/checkpoint_result.py (hunk 2 above)
Comment (Shinnnyshinshin) [reply great_expectations/checkpoint/types/checkpoint_result.py]: Thank you @alexsherstinsky, if that is the case, could I ask that individual tests for the methods be added in the next commit (or is it PR?)

**Review by Shinnnyshinshin — COMMENTED** [review]
Thank you very much for this work @alexsherstinsky. Looks good on my end too, with one clarification question

File: great_expectations/checkpoint/util.py [hunk 11]
```diff
@@ -236,6 +221,16 @@ def get_runtime_batch_request(
...
+    effective_batch_request: dict = dict(
+        **runtime_config_batch_request, **validation_batch_request
+    )
+    if "runtime_parameters" in effective_batch_request or isinstance(
+        validation_batch_request, RuntimeBatchRequest
+    ):
+        batch_request_class = RuntimeBatchRequest
+    else:
+        batch_request_class = BatchRequest
+
```
Comment (roblim) [inline great_expectations/checkpoint/util.py:234]: Much better!

File: great_expectations/checkpoint/types/checkpoint_result.py (hunk 9 above)
Comment (roblim) [reply great_expectations/checkpoint/types/checkpoint_result.py:292]: +1

**Comment by alexsherstinsky** [comment]
Tests have been added for new utilities.

In addition, `BatchRequest` serialization has also been dramatically simplified using these new utilities.

File: great_expectations/data_context/types/base.py [hunk 12]
```diff
@@ -1950,12 +1957,26 @@ def validate_schema(self, data, **kwargs):
...
+            config=obj, batch_data_references=batch_data_references
+        )
+        restore_runtime_parameters_batch_data_references_into_config(
+            config=data, batch_data_references=batch_data_references
+        )
+        return data
+
 
 class CheckpointConfig(BaseYamlConfig):
     # TODO: <Alex>ALEX (does not work yet)</Alex>
```
Comment (NathanFarmer) [inline great_expectations/data_context/types/base.py:1982]: Can we remove this?

**Review by NathanFarmer — COMMENTED** [review]
It is a thing of beauty. Just one question about a `TODO`.

File: great_expectations/data_context/types/base.py (hunk 12 above)
Comment (alexsherstinsky) [reply great_expectations/data_context/types/base.py:1982]: @NathanFarmer We still need those comments -- the idea was good, but it did not work, and we ought to figure out how to make it work -- or discuss it and conclude that the current approach is the best that can be done.  Thanks!

**Review by NathanFarmer — APPROVED** [review]
This is a wonderful refactor 😄 LGTM!

**Review by Shinnnyshinshin — APPROVED** [review]
Wonderful work 🙇🏼 thank you @alexsherstinsky

**Review by roblim — COMMENTED** [review]
Looks great! <3

File: great_expectations/core/batch.py [hunk 13]
```diff
@@ -751,11 +812,182 @@ def get_batch_request_dict(
...
+                    val["batch_request"]["runtime_parameters"]["batch_data"] = str(
+                        type(validations_batch_data_list[idx])
+                    )
+                else:
+                    val["batch_request"]["runtime_parameters"][
+                        "batch_data"
+                    ] = validations_batch_data_list[idx]
+
+
+def standardize_batch_request_display_ordering(
```
Comment (anthonyburdi) [inline great_expectations/core/batch.py:978]: 👍

File: great_expectations/core/usage_statistics/anonymizers/batch_request_anonymizer.py [hunk 14]
```diff
@@ -40,6 +41,9 @@ def anonymize_batch_request(self, *args, **kwargs) -> Dict[str, List[str]]:
         anonymized_batch_request_dict: Optional[
             Union[str, dict]
         ] = self._anonymize_batch_request_properties(source=batch_request_dict)
+        anonymized_batch_request_dict = standardize_batch_request_display_ordering(
```
Comment (anthonyburdi) [inline great_expectations/core/usage_statistics/anonymizers/batch_request_anonymizer.py:44]: 👍

File: tests/core/test_serialization.py [hunk 15]
```diff
@@ -59,3 +74,771 @@ def test_serialization_of_spark_df(spark_session):
     df = pd.DataFrame({"a": [1, 2, 3], "b": [4, 5, 6]})
     sdf = spark_session.createDataFrame(df)
     assert convert_to_json_serializable(sdf) == {"a": [1, 2, 3], "b": [4, 5, 6]}
+
+
+def test_batch_request_deepcopy():
```
Comment (anthonyburdi) [inline tests/core/test_serialization.py:79]: Thank you for adding these tests!

**Review by anthonyburdi — APPROVED** [review]
LGTM!
